python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -248,38 +250,76 @@ def run(self, mapperResults, exposure, **kwargs):
-                    subMI.getVariance().getArray()[isNotNan] = patchMI.getVariance().getArray()[isNotNan]
+                subMI.getImage().getArray()[isNotNan] = patchMI.getImage().getArray()[isNotNan]
+                subMI.getVariance().getArray()[isNotNan] = patchMI.getVariance().getArray()[isNotNan]
 
             if reduceOp == 'sum' or reduceOp == 'average':  # much of these two options is the same
-                if noNans:
-                    subMI += patchMI
-                    if reduceOp == 'average':
-                        # wsubim is a view into the `weights` Image, so here we simply add one to
-                        # the region of `weights` confined by `item.getBBox()`.
-                        wsubim = afwImage.ImageF(weights, item.getBBox())
-                        wsubim += 1.
-                else:
-                    isNotNan = ~isNan
-                    subMI.getImage().getArray()[isNotNan] += patchMI.getImage().getArray()[isNotNan]
-                    subMI.getVariance().getArray()[isNotNan] += patchMI.getVariance().getArray()[isNotNan]
-                    if reduceOp == 'average':
-                        wsubim = afwImage.ImageF(weights, item.getBBox())
-                        wsubim.getArray()[isNotNan] += 1.
+                subMI.getImage().getArray()[isNotNan] += patchMI.getImage().getArray()[isNotNan]
+                subMI.getVariance().getArray()[isNotNan] += patchMI.getVariance().getArray()[isNotNan]
+                if reduceOp == 'average':
+                    # wsubim is a view into the `weights` Image
+                    wsubim = afwImage.ImageI(weights, item.getBBox())
+                    wsubim.getArray()[isNotNan] += 1
 
         if reduceOp == 'average':
-            newMI /= weights
+            wts = weights.getArray().astype(np.float)
+            newMI.getImage().getArray()[:, :] /= wts
+            newMI.getVariance().getArray()[:, :] /= wts
+
+        # Not sure how to construct a PSF when reduceOp=='copy'...
+        if reduceOp == 'sum' or reduceOp == 'average':
+            psf = self._constructPsf(mapperResults, exposure)
+            newExp.setPsf(psf)
 
         return pipeBase.Struct(exposure=newExp)
 
+    def _constructPsf(self, mapperResults, exposure):
+        """Construct a CoaddPsf based on PSFs from individual subExposures
+
+        Currently uses (and returns) a CoaddPsf. TBD if we want to
+        create a custom subclass of CoaddPsf to differentiate it.
+
+        Parameters
+        ----------
+        mapperResults : list
+            list of `pipeBase.Struct` returned by `ImageMapperSubtask.run`.
+            For this to work, each element of `mapperResults` must contain
+            a `subExposure` element, from which the component Psfs are
+            extracted (thus the reducerTask cannot have
+            `reduceOperation = 'none'`.
+        exposure : afwImage.Exposure
```

Comment:
Will do in my commit to DM-7611, if that's okay (works off the same codebase and I don't want to add more conflicts).